web_src/js/features/aria.js

```diff
@@ -96,6 +96,11 @@ function attachOneDropdownAria($dropdown) {
   $dropdown.on('keyup', (e) => { if (e.key.startsWith('Arrow')) deferredRefreshAria(); });
 }
 
+export function initCancelButtons(rootElement=document) {
+  // Only cancel buttons should have the 'cancel' class, and these buttons should not submit an underlying form, so set 'type="button"' for them
+  rootElement.querySelectorAll('.cancel').forEach(cancelButton => cancelButton.setAttribute('type', 'button'));
+}
+
```

Comment:
Do we really need this? 🤔   

I think I fully understand the problem now.

* For Fomantic Dialog:  Maybe Fomantic UI doesn't do `preventDefault` for the `cancel button`, so we need the `type=button` on it if we decided to use `<button>`, or use `<div role=button tabindex=0>` to show that we do not want a form button.
* For other buttons: `initGlobalButtons` would `preventDefault` correctly, so these buttons do not need `type`.

Reply:
The good thing is: I can give your words directly back to you now :grin:
We can't ensure that we will forever stay with Fomantic UI (we seem to be on a really slow path to migrating to tailwind in the last few months).
Furthermore, we can't expect future developers to know exactly under which circumstances their buttons will misbehave.
So, with this change, they certainly know that all `cancel` buttons automatically behave correctly, no matter if Fomantic or we do some black magic or not.

> ps: `initCancelButtons` doesn't work for dynamically generated buttons (eg: AJAX?) in the future.

That's why I gave it a `rootElement` parameter. That way, even dynamically generated cancel buttons will be able to be converted.
Furthermore, I'm wondering under what circumstances we need to dynamically generate a cancel button?

Reply:
> Furthermore, I'm wondering under what circumstances we need to dynamically generate a cancel button?

Partially reloaded page? I can see there will be more and more AJAX requests in the future.

> We can't ensure that we will forever stay with Fomantic UI (we seem to be on a really slow path to migrating to tailwind in the last few months).

I just realized that this problem is only for Fomantic UI, because Gitea misued Fomantic UI modal.  Fomantic UI modal never expects the "ok" button to submit a form, never expects the "cancel" button is a form buton. So it doesn't do "preventDefault" on "cancel button"

Now, Gitea mixes everything together, so developers need to pay attention for the "form button" problem from now on ....

So I think maybe we don't need a "general cancel fix (initCancelButtons)"  to add anything.

> Furthermore, we can't expect future developers to know exactly under which circumstances their buttons will misbehave.

I think just like other `.show-panel.button`, we can do `$.on('.ui.modal .actions .cancel.button') { preventDefault }`, it's general enough if needed.

---

The whole picture (I think it could be the final one in my mind, sorry for the long discussion, it's just for reference & information):

* For our buttons like `.show-panel`, we just add `prevnetDefault`, that's good enough.
* Using "cancel button" in Fomantic UI modal with form needs to prevent the "submit action".
* JS `initCancelButtons` and `$.on('.ui.modal .actions .cancel.button')` all affect performance.
* The solutions for "cancel button" in Fomantic UI modal could be either:
    1. use `<button type=button>` or `<div role=button tabindex=0>`
    2. use `$.on('.ui.modal .actions .cancel.button')`  to catch all events
    3. use `initCancelButtons`

I think the first solution could be the best, no performance affect, it needs documents and comments.

The second solution could be general enough, newly created buttons could also be handled by it.

The third one works, but I prefer the first or the second.